multicluster/controllers/multicluster/member/resourceimport_controller.go

```diff
@@ -98,6 +101,8 @@ func newResourceImportReconciler(client client.Client, scheme *runtime.Scheme, l
 // Reconcile will attempt to ensure that the imported Resource is installed in local cluster as per the
 // ResourceImport object.
 func (r *ResourceImportReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.mutex.RLock()
```

Comment:
The `Reconcile` will be triggered by every request (maybe merged events), but each request is independent, is that possible that stale_controller.go get the mutex between two requests?

Reply:
It is possible I think. When `stale_controller` tries to lock, no active `ResourceImportReconciler` worker should be reconciling because of the `mutexassert` check. But in case after the check and immediately before the `stale_controller` locks the mutex, some ResourceImport starts reconciling, then the `stale_controller` will still have to wait. But the important thing is we make sure that when `stale_controller` is doing the cleanup, no concurrent ResourceImport reconciling should happen.

Reply:
StaleController reads ResourceImports from client cache right. It is possible that the cache is not up to date?

Do not we have a way to clean up stale resources first, before starting ResourceImportReconciler?

Reply:
> StaleController reads ResourceImports from client cache right. It is possible that the cache is not up to date?

Not if it is protected by mutex. The stale controller will take the lock to make sure no concurrent ResourceImport reconciling is in flight. ResourceImport cache will stay in place until it finishes.

> Do not we have a way to clean up stale resources first, before starting ResourceImportReconciler?

Lan Grayson and I discussed this offline. The rough idea is that we think it's better to have stale_controller back off as long as there's ResourceImportReconciler workers running, since the stale cleanup is of a lower priority than, for example, massive label identity sync during mc controller restart.

Reply:
I meant the K8s client cache, but I assume StaleController and ResourceImportReconciler should share the same client cache and read ResourceImports with the lock, so probably not an issue?

Reply:
Yes exactly.

Reply:
Ok. Still feel more deterministic to cleaning up stale resources first, but I am fine with the current approach too.